riptable/rt_dataset.py

```diff
@@ -2548,11 +2548,11 @@ def _to_unicode_if_string(arr):
                 categories = _to_unicode_if_string(col.category_array) if unicode else col.category_array
                 data[key]: pd.Categorical = pd.Categorical.from_codes(codes, categories=categories)
             elif isinstance(col, TypeRegister.DateTimeNano):
-                ccol = col.copy()
-                arr = ccol._timezone.fix_dst(ccol._fa)
-                arr = arr.astype('datetime64[ns]')
+                # N.B. Have to convert to numpy array because there is a bug in copying DataTimeNano

```

Comment:
Could you provide bug link for reference?

Reply:
I don't have a link but you can easily reproduce the problem with
```
pd.Series(pd.DatetimeIndex(rt.DateTimeNano(['19900401 02:00:00'], from_tz='NYC')))
```
I think it's also related to the internal working of pandas Series.

I just found that calling `view` on the DateTimeNano will trigger similar crash:
```
In [26]: rt.DateTimeNano(['19900401 02:00:00'], from_tz='NYC').view('i8')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/miniconda3/envs/riptable/lib/python3.8/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/miniconda3/envs/riptable/lib/python3.8/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395
    396             return _default_pprint(obj, self, cycle)

~/miniconda3/envs/riptable/lib/python3.8/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/miniconda3/envs/riptable/lib/python3.8/site-packages/riptable/rt_datetime.py in __repr__(self, verbose)
   3868     def __repr__(self, verbose=False):
   3869         repr_strings = []
-> 3870         tz_string = f", to_tz='{self._timezone._to_tz}'"
   3871         repr_strings.append(self.get_classname() + "([" + self._build_string() + "]" + tz_string + ")")
   3872

```

Reply:
Thank you, I've created a new issue for this (#206 )

Reply:
I have a fix for the DateTimeNano view casting bug at PR#207

Reply:
I tried with the patch but it has another error:
```
    @classmethod
    def newclassfrominstance(cls, instance, origin):
        '''
        Restore timezone/length info.
        '''
        result = instance.view(cls)
        result._timezone = origin._timezone.copy()
>       result._display_length = origin._display_length
E       AttributeError: 'DateTimeNano' object has no attribute '_display_length'
```
I think it boils down to creating pandas Series, this creates the same error:
```
pd.Series(pd.DatetimeIndex(rt.DateTimeNano(['19900401 02:00:00'], from_tz='NYC')))
```
Feels like DateTimeNano should have better design for creating new instance, otherwise if somebody adds a new attribute it must be added in multiple places to prevent blowup like this.

Reply:
Agree. This is another attribute that hasn't been copied as part of the view() call. I am going to write a test that verifies that a view of one of these subclass types contains all the same attributes as the original in an effort to find all that need to be copied.

Reply:
I switch to use the DateTimeNano array directly for constructing DatetimeIndex. Thanks.